[ResponseOps] use Data Streams for AAD indices in serverless

pmuellr opened the pull request:
resolves https://github.com/elastic/kibana/issues/154266

Changes the way the alerts-as-data (AAD) indices are created and written to, to allow them to be built as they have been in the past (alias and backing indices created manually) OR as an ES Data Stream.

Serverless will use Data Streams, other environments will use the existing alias and backing indices. The determination is made by optionally including the `serverless` plugin, and determining if it's available.

The implementation is organized around a `DataStreamAdapter` object, which is instantiated with a "data stream" or "alias" flavor, and then it handles the specialized behavior.  Currently, a lot of the smaller implementation bits, like setting property values in ES calls, is done via in-line boolean checks of that object, as to whether data streams or aliases are being used.  This could probably be cleaned up some.

Existing non-serverless function tests are largely unchanged, as they can't test the new data stream path.  Some tests have been added to the serverless function tests, to test basic reading / writing via updated alert documents.

## DEFER

- more serverless AaD tests

- potentially re-org `alerts_service` tests; currently this module makes no direct ES calls, but it's jest tests are quite large and mock a LOT of ES calls.  The calls are now indirect in helpers.  Almost feels like the code was re-org'd to add the helpers, but the tests weren't.  Would likely make the tests more understandable with a re-org - spent lots of time in this test module
  - x-pack/plugins/alerting/server/alerts_service/alerts_service.test.ts
  - x-pack/plugins/alerting/server/alerts_service/lib/create_or_update_ilm_policy.test.ts
  - x-pack/plugins/alerting/server/alerts_service/lib/create_or_update_index_template.test.ts
  - x-pack/plugins/alerting/server/alerts_service/lib/data_stream_adapter.ts

- https://github.com/elastic/kibana/issues/158403 - this issue is more noticeable now that we HAVE to do OCC with data streams, so we get errors instead of simply overwriting documents (which is also bad)

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

pmuellr committed:
[ResponseOps] use Data Streams for AAD indices in serverless

resolves https://github.com/elastic/kibana/issues/154266

Changes the way the alerts-as-data (AAD) indices are created and
written to, to allow them to be built as they have been in the past
(alias and backing indices created manually) OR as an ES Data Stream.

Serverless will use Data Streams, other environments will use the
existing alias and backing indices.  The determination is made by
optionally including the `serverless` plugin, and determining if it's
available.

mikecote commented on x-pack/plugins/alerting/server/plugin.ts:
```diff
@@ -229,6 +235,14 @@ export class AlertingPlugin {
     this.licenseState = new LicenseState(plugins.licensing.license$);
     this.security = plugins.security;
 
+    const useDataStreamForAlerts = !!plugins.serverless;
```
**question:** Did we want this determined by the existence of the serverless plugin or via the `xpack.alerting.useDataStreamForAlerts` yml setting? We have the ability to set serverless specific configs inside `config/serverless.yml`.

pmuellr replied:
I did originally have a config - may need to remove some remaining bits of it.

To do it right, you're supposed to make it conditional on using the `--serverless` (or whatever) option, so that in theory it can't be used in non-serverless.  In the end, to keep it "safe" (we can't use DS for non-serverless, nor aliases for serverless) for now, basing it off the serverless plugin was easier, but also means we can only do functional / integration testing in the serverless tests.

I also realized lots of our existing FT can't be used directly in serverless, as they do things like mess with spaces, which isn't fully supported in serverless.  If we _could_ have (I thought we might), I might have leaned into the config a bit more, as we could run some of our FT with datastreams, some with alias.  Nope!

I think we could change the way we do this later, if we decide we want it more explicit, without breaking any public "API" (since there is none now).

mikecote replied:
This all makes sense to me, thank you! 👍

pmuellr committed:
finish jest tests

pmuellr committed:
fix a test, revert some error behavior per PR comment

pmuellr committed:
merge main

pmuellr committed:
fix up FAaD alert read/write, add serverless test

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

mikecote commented:
All I have left is to pull down locally and test, the rest looks good! I will finish my review soon 👍

pmuellr commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.test.ts:
```diff
@@ -81,406 +82,310 @@ const mockSetContext = jest.fn();
-      expect(mockLegacyAlertsClient.getTrackedAlerts).not.toHaveBeenCalled();
-      spy.mockRestore();
-    });
+  for (const useDataStreamForAlerts of [false, true]) {
```
A lot of the changes to jest tests are similar to this.  They now run the existing tests twice - once for not using data streams for alerts, once for using data streams as alerts.  You'll want "hide whitespace changes" on for these, though it seems like it's on by default for me now?  \o/

pmuellr commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.test.ts:
```diff
@@ -81,406 +82,310 @@ const mockSetContext = jest.fn();
-                flapping_history: [true],
-                instance: {
-                  id: '2',
+            index: useDataStreamForAlerts
```
There are a lot of snippets like this in the tests - inline testing of whether data streams are being used.

I don't feel like I had a good "view" of this while hacking away, fixing these.  Perhaps a lot of these can be re-org'd somehome into some simple functions ... as a later enhancement.

pmuellr commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.test.ts:
```diff
@@ -720,40 +467,157 @@ describe('Alerts Client', () => {
+                {
+                  _id: 'abc',
+                  _index: '.internal.alerts-test.alerts-default-000001',
+                  _seq_no: 41,
```
Here we see one of the changes to existing behavior.  When querying for previous alerts, we also get the seqNo/primaryTerm info via some new option.  We need these for optimistic concurrency control (OCC) updates to alert docs, with data streams.  Previously, we could just bulk `index` updates to alerts, now we need to provide OCC info as well ...

pmuellr commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.test.ts:
```diff
@@ -770,1217 +634,1371 @@ describe('Alerts Client', () => {
-                instance: {
-                  id: '2',
+              {
+                create: { _id: uuid2, ...(useDataStreamForAlerts ? {} : { require_alias: true }) },
```
One general change that was made was that for **new** alerts, we use `create` instead of `index`, since we can't use `index` to create new docs in data streams.  We continue to use `index` to **update** alert docs, but now need OCC info for that as well.

pmuellr commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.ts:
```diff
@@ -78,6 +80,8 @@ export class AlertsClient<
   private fetchedAlerts: {
     indices: Record<string, string>;
     data: Record<string, Alert & AlertData>;
+    seqNo: Record<string, number | undefined>;
```
Here is where we capture the seqNo/primaryTerm for the previous alert docs, in case we need to update them.

pmuellr commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.ts:
```diff
@@ -131,6 +137,7 @@ export class AlertsClient<
     const queryByUuid = async (uuids: string[]) => {
       const result = await this.search({
         size: uuids.length,
+        seq_no_primary_term: true,
```
This is how seqNo/primaryTerm end up showing up in the result

pmuellr commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.ts:
```diff
@@ -174,11 +183,15 @@ export class AlertsClient<
-      index: this.indexTemplateAndPattern.pattern,
+      index,
       body: queryBody,
+      ignore_unavailable: true,
```
A new thing!  While comparing ruleRegistry to this code, I noticed they were doing `ignore_unavailable: true`, which seems like it makes sense for us as well.  Basically, drive-by.  No problem removing, but seems useful.

pmuellr commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.ts:
```diff
@@ -360,34 +373,31 @@ export class AlertsClient<
         if (response && response.errors) {
           const errorsInResponse = (response.items ?? [])
-            .map((item) => (item && item.index && item.index.error ? item.index.error : null))
+            .map((item) => item?.index?.error || item?.create?.error)
```
simplification of existing code, and we need to check for `create` errors as well

pmuellr commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.ts:
```diff
@@ -402,6 +412,33 @@ export class AlertsClient<
       }
     }
+
+    function getBulkMeta(
```
Here are the main changes to how we "write" documents, depending on create/update.  Updates are via `index` with OCC fields applied.  Creating new docs is via `create`.  This is the way (of data streams).

Should note here, this is a behavior change.  Previously, we just did `index`, which for alias/indices just overwrote the existing document.  So in race conditions, the second one won.  Now, with OCC, the second one will lose.  I figure, doesn't matter which loses.  Though we may see some specific patterns where this behavior change **DOES** matter.  

We need to fix this problem somehow, and the PR points to a specific GH issue surrounding this.

pmuellr committed:
Merge branch 'main' into alerting/serverless-aad-to-datastream

fix conflicts in:

- x-pack/plugins/alerting/server/alerts_client/alerts_client.test.ts
- x-pack/plugins/alerting/server/plugin.test.ts
- x-pack/plugins/alerting/server/plugin.ts
- x-pack/plugins/alerting/server/task_runner/task_runner_alerts_client.test.ts

pmuellr commented on x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts:
```diff
@@ -230,10 +234,18 @@ export const createLifecycleExecutor =
-            fields: hit._source,
-          };
+          if (hit._seq_no == null) {
+            logger.error(`missing _seq_no on alert instance ${alertInstanceId}`);
+          } else if (hit._primary_term == null) {
+            logger.error(`missing _primary_term on alert instance ${alertInstanceId}`);
```
good idea; seems extremely unlikely we'd see this, it's more of a check that we are actually copying the values from the fetched docs, to use when we need to update them later.

pmuellr committed:
update logged error messages for seq_no/primary_term per PR request

pmuellr committed:
[EventLog] don't use legacy template APIs during initialization

resolves https://github.com/elastic/kibana/issues/164731

This removes the check for a legacy index template for the event log
for the current version it wants to create.  This was presumably done
this way, to allow for the switch from legacy to component templates.

But it's not really needed, since each stack version creates new index
templates.  So, removing the check should be fine.

pmuellr commented:
I just pushed commit 943630edfdc2626637fe5e79ed6ec19f0c369671 into this PR, which fixes [a different serverless issue we found in event log](https://github.com/elastic/kibana/pull/164747).  I'm having problems getting CI to allow me to merge it.  So, merging it here for now, that PR should be merged soon.

pmuellr committed:
fix serverless function test

pmuellr committed:
Merge branch 'main' into alerting/serverless-aad-to-datastream

fix conflicts in

- x-pack/plugins/security_solution/server/lib/risk_engine/risk_engine_data_client.test.ts
- x-pack/test_serverless/shared/lib/index.ts

pmuellr committed:
fix a flaky serverless test, fix risk engine to pass ilm-y dataStreamAdapter

pmuellr committed:
fix more flakiness

pmuellr committed:
add more un-flay-ness to FT

pmuellr commented:
Some details on the serverless security cypress failures:

- they are not blocking merging; the servless tests are "soft fails"
- the non-flaky issues here are all around the alerts indices - either creating them, or writing to them
- there could be further errors in these tests after these initial startup issues in the tests are resolved

----

**7/16 - Serverless Security Cypress Tests**

The problem is esArchiver is being used to re-create some alert indices, but now it can't since the data stream exists.  I think we can just change this test to use `docs_only`, so that it won't try to recreate the indices, just use what's there.

source: x-pack/test/security_solution_cypress/cypress/e2e/detection_alerts/ransomware_detection.cy.ts

----

**8/16 - Serverless Security Cypress Tests**

same problem, same file as 7/16

----

**1/6 - Serverless Security Investigations Cypress Tests**

Data stream problem, but different.  This time it seems to be complaining about using bulk op `index` to create docs, when we need to `create` instead, for data streams.  Another likely change, will be that it's likely using `require_alias: true`, but that won't work with data streams as we don't have an alias anymore, just the data stream.

source: x-pack/test/security_solution_cypress/cypress/e2e/investigations/alerts/alert_table_action_column.cy.ts

----

**2/6 Serverless Security Investigations Cypress Tests**

Same as 7/16 - Serverless Security Cypress Tests

source: x-pack/test/security_solution_cypress/cypress/e2e/investigations/alerts/alerts_details.cy.ts

patrykkopycinski committed:
add support for options in EsArchiver.load

mikecote reviewed (approved):
Tested locally and reviewed the code, LGTM! 👍

patrykkopycinski committed:
fix update alert status

pmuellr committed:
Revert "fix update alert status"

This reverts commit c9d53e0f3f85e7689ad4e4340fd12c5930efc71c.

Not clear this fix was needed, as one of the tests in
alerts_details.cy.ts is still failing; this also broke
jest and FTR tests.

patrykkopycinski committed:
Merge branch 'main' of github.com:elastic/kibana into alerting/serverless-aad-to-datastream

# Conflicts:
#	x-pack/test/security_solution_cypress/cypress/e2e/detection_alerts/ransomware_detection.cy.ts
#	x-pack/test/security_solution_cypress/cypress/e2e/detection_alerts/ransomware_prevention.cy.ts
#	x-pack/test/security_solution_cypress/cypress/e2e/investigations/alerts/alert_table_action_column.cy.ts
#	x-pack/test/security_solution_cypress/cypress/e2e/investigations/alerts/alerts_details.cy.ts

patrykkopycinski committed:
skip on serverless

patrykkopycinski committed:
Revert "Revert "fix update alert status""

This reverts commit c8b065a8e962c7e37254e256890cca6dcd61f2d3.

patrykkopycinski committed:
fix update alert status

patrykkopycinski committed:
unskip

patrykkopycinski committed:
Merge branch 'main' into alerting/serverless-aad-to-datastream

pmuellr committed:
merge main and fix conflicts

pmuellr commented:
@elasticmachine merge upstream

I think this should clear the failing FTR test via PR 165103

kibanamachine committed:
Merge branch 'main' into alerting/serverless-aad-to-datastream

pmuellr committed:
merge main, fix conflicts

MadameSheema reviewed (approved):
security-engineering-productivity changes LGTM

PhilippeOberti reviewed (approved):
LGTM for the Threat Hunting Investigations team

patrykkopycinski committed:
Merge branch 'main' into alerting/serverless-aad-to-datastream

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/signals/open_close_signals_route.ts:
```diff
@@ -121,17 +121,22 @@ const updateSignalsStatusByIds = async (
+    // https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update-by-query.html#_refreshing_shards_2
+    // Note: Before we tried to use "refresh: wait_for" but I do not think that was available and instead it defaulted to "refresh: true"
+    // but the tests do not pass with "refresh: false". If at some point a "refresh: wait_for" is implemented, we should use that instead.
+    refresh: true,
```
is allowing `refresh:true` something we want to do?
I remember we did some work to avoid explicitly refreshing indices: https://github.com/elastic/security-team/issues/6561

vitaliidm replied:
In similar work, done in https://github.com/elastic/kibana/pull/162508, I had the same issue, so had add retry calls to wait until item reindexed and only after this finish request.

It worth to note though, updates in lists/items are not available in UI, only through API calls, so we do not expect them to be called frequently, especially in Serverless. Only list delete available in UI, which is also not expected to be frequent operation

patrykkopycinski replied:
Thank you for the review @vitaliidm, I took it from here https://github.com/elastic/kibana/blob/4d6f73257487d3ddd86905a7e0030061524ca639/x-pack/plugins/security_solution/server/lib/detection_engine/routes/signals/open_close_signals_route.ts#L130

patrykkopycinski replied:
I have tested it locally and it seems `false` works fine, let's see if CI agrees 🤞

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/signals/set_alert_tags_route.ts:
```diff
@@ -90,9 +90,20 @@ export const setAlertTagsRoute = (router: SecuritySolutionPluginRouter) => {
+          // https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update-by-query.html#_refreshing_shards_2
+          // Note: Before we tried to use "refresh: wait_for" but I do not think that was available and instead it defaulted to "refresh: true"
+          // but the tests do not pass with "refresh: false". If at some point a "refresh: wait_for" is implemented, we should use that instead.
+          refresh: true,
```
same as above

> is allowing refresh:true something we want to do?
I remember we did some work to avoid explicitly refreshing indices: https://github.com/elastic/security-team/issues/6561

e40pud reviewed (approved):
DE changes LGTM

patrykkopycinski committed:
cr

patrykkopycinski committed:
skip tests

patrykkopycinski committed:
skip
